### DISCUSSION TITLE
Offloading Yarp crypto operations to OpenSSL Engine. Is that possible?

### DISCUSSION THREAD

**Discussion by JEduCB** [dotnet/runtime#79714] [Q&A] [answered]
Assuming I have a well-configured OpenSSL Engine in my Linux Ubuntu 20.04.5 system - `openssl engine` and `openssl ...` commands (and the engine logs) show me its available and working.

And this is my `/etc/ssl/openssl.conf` config file (based on [this comment](https://github.com/dotnet/runtime/issues/36938#issuecomment-651934082)):

```
openssl_conf = openssl_init

[ openssl_init ]
engines = engine_section

[ engine_section ]
X = X_section

[ X_section ]
engine_id = X_engine
dynamic_path = /usr/lib/x86_64-linux-gnu/engines-1.1/X_engine.so

default_algorithms = ALL
init = 1
```

So, considering the engine is working, if I implement a HTTPS proxy server using Yarp, may I assume the .net/OpenSSL stack would automatically offload crypto requests to the OpenSSL Engine?

If not, how can I configure Yarp to do that?
Could I do that by config files (like `appsettings.json` i.e.) or would it required additional code to load the engine dynamically?
Is that offloading even possible?

Thanks

`# cat /etc/os-release`
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"

`# dotnet --list-sdks`
3.1.425 [/usr/share/dotnet/sdk]
6.0.403 [/usr/share/dotnet/sdk]
7.0.100 [/usr/share/dotnet/sdk]

`# dotnet --list-runtimes`
Microsoft.AspNetCore.App 3.1.31 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 6.0.11 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 7.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.NETCore.App 3.1.31 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 6.0.11 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
Microsoft.NETCore.App 7.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

**Comment by teo-tsirpanis** [comment, marked as answer]
.NET on Linux will automatically use OpenSSL for its cryptograpy algorithms. I don't think you have to do anything special.

**Reply by JEduCB** [reply]
Hi @teo-tsirpanis,

Right, but for some reason the engine was not working. After looking into the `LD_DEBUG=libs` output I noticed both `libssl.so.1.1` and the engine lib were not being properly loaded.

Solved including both libs in `LD_PRELOAD` env var.

Thanks for the comment.

Cheers.
